src/elements/train/train-wagon/train-wagon.ts

```diff
@@ -90,126 +101,137 @@ class SbbTrainWagonElement extends SbbNamedSlotListMixin<SbbIconElement, typeof
-
-    if (
-      changedProperties.has('type') ||
-      changedProperties.has('occupancy') ||
-      changedProperties.has('wagonClass')
-    ) {
-      this.toggleAttribute(
-        'data-has-visible-wagon-content',
-        Boolean((this.type === 'wagon' && this.occupancy) || this.wagonClass),
-      );
-    }
-  }
-
   private _sectorChanged(): void {
     this._sectorChange.emit();
   }
 
   protected override render(): TemplateResult {
-    const label = (tagName: string): TemplateResult => {
-      const TAG_NAME = tagName;
-      /* eslint-disable lit/binding-positions */
-      return html`
-        <${unsafeStatic(TAG_NAME)} class="sbb-train-wagon__label" aria-hidden=${(!this
-          .label).toString()}>
-          ${
-            this.label
-              ? html` <span class="sbb-screen-reader-only">
-                    ${`${i18nWagonLabelNumber[this._language.current]},`}&nbsp;
-                  </span>
-                  ${this.label}`
-              : nothing
-          }
-        </${unsafeStatic(TAG_NAME)}>
-      `;
+    const typeLabel = (): string => {
+      if (this.type === 'closed') {
+        return i18nClosedCompartmentLabel(this.label ? parseInt(this.label) : undefined)[
+          this._language.current
+        ];
+      } else if (this.type === 'locomotive') {
+        return i18nLocomotiveLabel[this._language.current];
+      } else if (this.type === 'couchette') {
+        return i18nCouchetteWagonLabel[this._language.current];
+      } else if (this.type === 'sleeping') {
+        return i18nSleepingWagonLabel[this._language.current];
+      } else if (this.type === 'restaurant') {
+        return i18nRestaurantWagonLabel[this._language.current];
+      }
+      return i18nWagonLabel[this._language.current];
     };
 
+    // From accessibility perspective we cannot create a list with only one entry.
+    // We need to know what will be presented and switch semantics based on count.
+    const listEntriesCount =
+      +!!this.sector +
+      +(this.label && this.type !== 'closed') +
+      +(!!this.wagonClass && this.type !== 'closed') +
+      +(!!this.occupancy && this.type !== 'closed') +
+      +(this.blockedPassage && this.blockedPassage !== 'none');
+
     const sectorString = `${i18nSector[this._language.current]}, ${this.sector}`;
 
+    const labelContent =
+      this.label && this.type !== 'closed'
+        ? html`<span class="sbb-screen-reader-only">
+              ${`${i18nWagonLabelNumber[this._language.current]}, ${this.label}`}
+            </span>
+            <span aria-hidden="true">${this.label}</span>`
+        : nothing;
+
+    const wagonClassContent = html`<span class="sbb-screen-reader-only">
+        ${this.wagonClass === '1'
+          ? i18nClass['first'][this._language.current]
+          : i18nClass['second'][this._language.current]}
+      </span>
+      <span aria-hidden="true">${this.wagonClass}</span>`;
+
+    const mainIcon = choose(this.type, [
```

Comment:
Nitpick; From my understanding, the only thing that changes is the icon name. Could this not be solved by just mapping the type to the icon name and setting the attribute accordingly?